tools/yulPhaser/Population.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/Population.h>
+
+#include <tools/yulPhaser/Program.h>
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <numeric>
+
+using namespace std;
+using namespace solidity::langutil;
+
+namespace solidity::phaser
+{
+
+size_t Population::measureFitness(Chromosome const& _chromosome, CharStream& _sourceCode)
+{
+	auto program = Program::load(_sourceCode);
+	program.optimise(_chromosome.optimisationSteps());
+	return program.codeSize();
+}
+
+ostream& operator<<(ostream& _stream, Individual const& _individual)
+{
+	_stream << "Fitness: ";
+	if (_individual.fitness.has_value())
+		_stream << _individual.fitness.value();
+	else
+		_stream << "<NONE>";
+	_stream << ", optimisations: " << _individual.chromosome;
+
+	return _stream;
+}
+
+Population::Population(CharStream _sourceCode, vector<Chromosome> const& _chromosomes):
+	m_sourceCode{std::move(_sourceCode)}
+{
+	for (auto const& chromosome: _chromosomes)
+		m_individuals.push_back(Individual{chromosome, nullopt});
+}
+
+Population Population::makeRandom(CharStream _sourceCode, size_t _size)
+{
+	vector<Individual> individuals;
+	for (size_t i = 0; i < _size; ++i)
+		individuals.push_back(Individual{
+			Chromosome::makeRandom(randomChromosomeLength()),
+			nullopt
+		});
+
+	return Population(move(_sourceCode), individuals);
+}
+
+void Population::run(optional<size_t> _numRounds, std::ostream& _outputStream)
+{
+	assert(_numRounds == nullopt || _numRounds >= 0);
+
+	doEvaluation();
+	for (size_t round = 0; !_numRounds.has_value() || round < _numRounds.value(); ++round)
+	{
+		doMutation();
```

Comment:
Shouldn't it be mutation - evaluation - selection?

Reply:
Yes, but I wanted to see the fitness for all chromosomes in the output and didn't really want to run evaluate twice :) So I put it after selection but this forced me to do one evaluation call before the loop.

It's not great but, but this code will change completely anyway in the next PR so I'd leave it as is here. See my explanation in https://github.com/ethereum/solidity/pull/8164#discussion_r371950090